src/metabase/models/cloud_migration.clj

```diff
@@ -0,0 +1,213 @@
+(ns metabase.models.cloud-migration
+  "A model representing a migration to cloud."
+  (:require
+   [cheshire.core :as json]
+   [clj-http.client :as http]
+   [clojure.java.io :as io]
+   [clojure.set :as set]
+   [metabase.cmd.dump-to-h2 :as dump-to-h2]
+   [metabase.models.interface :as mi]
+   [metabase.models.setting :refer [defsetting]]
+   [metabase.models.setting.cache :as setting.cache]
+   [metabase.util.i18n :refer [deferred-tru tru]]
+   [metabase.util.log :as log]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.pipeline :as t2.pipeline])
+  (:import
+   [java.io File InputStream]))
+
+(set! *warn-on-reflection* true)
+
+(def CloudMigration "Cloud Migration" :model/CloudMigration)
+
+(doto :model/CloudMigration
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(methodical/defmethod t2/table-name :model/CloudMigration [_model] :cloud_migration)
+
+(t2/deftransforms :model/CloudMigration
+  {:state mi/transform-keyword})
+
+(defsetting metabase-store-migration-url
+  (deferred-tru "Store URL for migrations. Internal test use only.")
+  :visibility :internal
+  :default    "https://store-api.metabase.com/api/v2/migration"
+  :doc        false
+  :export?    false)
+
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
+         "Audit tables are excluded from read-only-mode mode."))
+  :type       :boolean
+  :visibility :admin
+  :default    false
+  :doc        false
+  :export?    false)
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    (t2/table-name :model/CloudMigration)
+    (t2/table-name :model/Setting)
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    (t2/table-name :model/User)
+    (t2/table-name :model/Session)
+    (t2/table-name :model/LoginHistory)
+    (t2/table-name :model/AuditLog)
+    (t2/table-name :model/QueryExecution)
+    (t2/table-name :model/ViewLog)})
+
+(def ^:private ^:dynamic
+  *ignore-read-only-mode*
+  "Used during dump-to-h2, since rotate-encryption-key! over the dump will hit read-only-mode."
+  nil)
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
+                                                 #_model          :default
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions (t2/table-name model)))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info (tru "Metabase is in read-only-mode mode!")
+                    {:status-code 403})))
```

Comment:
i get that "you don't have permission to access this resource" kinda makes sense here, but 403 has a pretty standard meaning and throwing it to logged in users feels kinda weird. Is there a better status code for this?

Reply:
Going down some candidates in https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses:
- 400: the request is fine I think
- 401: client is logged in already
- 403: (the current one) we know the client but they can't do this op (right now)
- 409: this one seems reasonable... it is state related
- 412: there's a precondition missing (read-only false) but not in the headers
- 417: there's a failed expectation (read-only false) but not in the Expect header

So I'd say either 403 or 409. Up to you.

Reply:
403 is intended to be used for logged in users that don't have the authorization to perform a specific operation - `Unlike 401 Unauthorized, the client's identity is known to the server. ` 
I don't think anything else in 4xx/5xx fits.